---
title: "02_plotting_2"
author: "JR"
date: "8/3/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(tidyverse)
library(GenomicRanges)
source("../../../util/plotting_functions.R")
source("../../../util/_setup.R")
library(Gviz)
library(ggpubr)
library(ggplot2)
```
Accompanying youtube video:
https://youtu.be/PWYfKUw3Xyw

Goal: learn more plotting skills using our custom ChIPseq data set!
We will start to make plots that present results about our DBPs binding features.

Let's start making plots of our data !

# Now we can simply load the .Rdata file from before !!
we saved our environment in 01_peak_features -- let's load it in.
```{r loading in consensus peaks}

load("../01_peak_features/results/peak_features.RData", verbose = T)
# that is so nice to have everything loaded in stead of remaking everything !

```

# Peaks per dbp
```{r plotting peak features}

# First let's look at a histogram of peak#/DBP

 ggplot(num_peaks_df, aes(x = num_peaks)) + 
  geom_histogram(bins = 10)

```


# plotting num_peaks versus genome coverage.
? What would you suspect would happen
```{r peaks vs coverage}

# Let's plot our result of num_peaks versus genome coverage.
ggplot(num_peaks_df, aes(x = num_peaks, y = total_peak_length)) +
  geom_point() + 

  ylab("BP covered") +
  xlab("Number of peaks") +
  ggtitle("Peak count vs. total bases covered")

# saving MAKE SURE WORKING DIRECTORY IS CORRECT
ggsave("figures/peak_num_vs_coverage.pdf")

```

# plotting num peaks on promoters
Let's did in a bit deeper and see how coverage on promoters.
```{r number of DBPS on promoters}

# now let's plot x as num_peaks and y num overlapping promoters.

ggplot(num_peaks_df,
       aes(x = num_peaks, y = peaks_overlapping_promoters)) +
  geom_point() +
  
  geom_abline(slope = 1, linetype="dashed") +
  geom_smooth(method = "lm", se=F, formula = 'y ~ x',
              color = "#a8404c") +
 # ggpubr prints equation
  stat_regline_equation()+
  # stat_regline_equation(label.x = 35000, label.y = 18000) +
  ylim(0,60100) +
  xlim(0,60100) +

  # adding labels
xlab("Peaks per DBP") +
  ylab("Number of peaks overlapping promoters") +
  ggtitle("Relationship Between Number of DBP Peaks and Promoter Overlaps")

ggsave("figures/3_peak_num_vs_promoter_coverage.pdf")

# we observe a "saturation of binding events -- as you get more
# peaks you stop increaseing binding to promoters.
# maybe it leaks into the gene body let's check
```


# peak Coverage on gene bodies
```{r peak coverage on gene bodies}

ggplot(num_peaks_df,
       aes(x = num_peaks, y = peaks_overlapping_genebody)) +
 geom_point() +
  
  geom_abline(slope = 1, linetype="dashed") +
  geom_smooth(method = "lm", se=F, formula = 'y ~ x',
              color = "#a8404c") +
  stat_regline_equation(label.x = 35000, label.y = 18000) +
  
  ylim(0,60100) +
  xlim(0,60100) +

  # adding labels
xlab("Peaks per DBP") +
  ylab("Number of peaks overlapping genes") +
  ggtitle("Relationship Between Number of DBP Peaks and Gene Body Overlaps")

# Interesting result !! Gene bodies explain almost all the places of binding in 
# the genome! Where as promoters had a non linear asymptope after a certain number 
# of peaks. VERY COOL and surprising. There are billions of other possible binding sites!

ggsave("figures/4_peak_num_vs_gene_body_coverage.pdf")
```

Ok so we have started generating results in this handy num_peaks_df. 
However, the peak_occurence_df is also very handy for analyses.
Let's load it in and do some analyses -- THIS ONE IS REALLY COOL :).

# Reading in Peak occurence df

```{r read in peak_occurence_df}

# First read the file into an object

peak_occurence_df <- read.csv("../01_peak_features/results/peak_occurence_dataframe.csv")

# peak_occurence_df <- read.csv("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/12_peak_features/peak_occurence_dataframe.csv")

# let's have a look. This is a summary of the peak occurence df.
# we did colsums to find how many binding events there are at each promoter
# so the last column will tell us how many binding events there were at a given promoter. 

```

Let's make a density plot of num DBPs bound per promoter
# Density plot of binding events

```{r density plot of DBP localization events}

# Since we are making a density plot we only need the X axis.

ggplot(peak_occurence_df, aes(x = number_of_dbp)) +
geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA and lncRNA genes") 

# saving
ggsave("figures/num_binding_events_per_promoter.pdf")

```


Lets find how many promoters don't have any DBPs bound
# promoters with out binding events

```{r prmoters with out binding events}

# here is just a simple index and filter of the index to have at least 1 dbp bound.
unbound_promoters <- peak_occurence_df %>% 
  filter(peak_occurence_df$number_of_dbp < 1)

# how many are there?
nrow(unbound_promoters)

#  let's put it in a folder called results. We will always use this folder structure
write_csv(unbound_promoters, "../01_peak_features/results/unbound_promoters.csv")

```


# Do any DBPs never bind a promoter?
```{r is there a DBP that doesn't bind promoters?}

# we actually don't need code for this we can look in num_peaks_df and sort :)
# There are only a few and we may want to filter these out above.

```

# lncRNA versus mRNA promoter binding
Let's compare the binding patterns of lncRNA vs mRNA promoters.
While we are at it let's do some deep dives into the aes layer
This is required to make two plots in one :)

```{r}
# Plotting lncRNA -vs mRNA promoter binding distribution
# for more advanced figure making check out the code below.
# it is not the best practice to do to many analyses in ggplot.
# The below esxample shows how powerful and flexible ggplot is.

ggplot(num_peaks_df, aes(x = num_peaks)) +
  geom_point(aes(y = peaks_overlapping_lncrna_promoters), color = "red") +
  geom_point(aes(y = peaks_overlapping_mrna_promoters), color = "black") +
  
  stat_regline_equation(aes(y = peaks_overlapping_lncrna_promoters), color = "red") +
  stat_regline_equation(aes(y = peaks_overlapping_mrna_promoters), color = "black", label.y = 20000) +
  
  geom_smooth(aes(y = peaks_overlapping_lncrna_promoters), method = "lm", se = FALSE, formula = "y ~ x") +
  geom_smooth(aes(y = peaks_overlapping_mrna_promoters), method = "lm", se = FALSE, formula = "y ~ x")

# saving
ggsave("figures/lncRNA-mRNA-promoter_binding.pdf")

# Nice we see that mRNA promoters have about 4x more binding (from slope)
# Both follow a fairly linear trend except mRNA may have saturate quicker.
```

Fancier version of num_peaks vs promoter overlaps
Using mutate to create gene-type of lncRNA and mRNA seperately
Then plotting.
```{r}

# removing col "peaks_overlapping_promoters" 
num_peaks_dfl <- num_peaks_df %>%
  dplyr::select(-peaks_overlapping_promoters) %>%
  
  # We will discuss pivot longer in the next class (BUT IF YOUR INTERESTED ASK NOW IT'S WORTH IT!!)
  pivot_longer(cols = peaks_overlapping_lncrna_promoters:peaks_overlapping_mrna_promoters,
               names_to = "gene_type",
               values_to = "peaks_overlapping_promoters") %>%
  mutate(gene_type = gsub("peaks_overlapping_", "", gene_type))

# plotting
ggplot(num_peaks_dfl, aes(x = num_peaks, y = peaks_overlapping_promoters, 
                         col = gene_type)) +
  geom_point() +
  
         geom_abline(slope = 1, linetype="dashed") +
  geom_smooth(method = "lm", se = FALSE, formula = "y ~ x") +
  
  stat_regline_equation() +
  
  scale_color_manual(values = c("#a8404c", "#424242"))+
  
  xlab("Peaks per DBP") +
  ylab("Peaks Overlapping Promoters") +
  ggtitle("Number of DBP Peaks and Promoter Overlaps")


#ggsave("figures/peaks_overlaps_relationship_by_gene_type.png", height = 5, width = 8)
ggsave("figures/peaks_overlaps_relationship_by_gene_type.pdf", height = 5, width = 8)
```

############
exeRcise:
############

Rerun the code above with big_num_peaks_df -- pick an 
interesting figure and present.

