---
title: "02_R_basics"
author: "JR"
date: "7/20/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(tidyverse)
library(GenomicRanges)
source("/scratch/Shares/rinnclass/CLASS_2023/JR/CLASS_2023/util/intersect_functions.R")
source("/scratch/Shares/rinnclass/CLASS_2023/JR/CLASS_2023/util/_setup.R")
```

Today we are going to do some basic R commands and look at our Pol2 ChIP data.
We will use read_tsv to load in a peak file that is in tab separated (tsv)

```{bash}

# The Peak files are located in our 03 nextflow folder -- let's take a look:

cd /scratch/Shares/rinnclass/CLASS_2023/<your_folder>/CLASS_2023/CLASSES/03_Nextflow/00_Pol2_NF_CORE_RUN/results/bwa/mergedLibrary/macs/broadPeak

# /scratch/Shares/rinnclass/CLASS_2023/JR/CLASS_2023/CLASSES/03_Nextflow/00_Pol2_NF_CORE_RUN/results/bwa/mergedLibrary/macs/broadPeak

```

We see a lot of different types of peak files and we want .braodPeak
Let's load one in 

```{r basics of reading in peak file}
# loading in peak file 1:
# we will first need to set an object name to load into the environment (what we are calling this table)
# we assign it the values in the file using the assignment opperator " <- " 
# then the file path to the peak file

Pol2_peak1 <- read_tsv("/scratch/Shares/rinnclass/CLASS_2023/JR/CLASS_2023/CLASSES/03_Nextflow/00_Pol2_NF_CORE_RUN/results/bwa/mergedLibrary/macs/broadPeak/POLR2A_R1_peaks.broadPeak")

# cool, let's take a look at what we got by clicking the object in the env window.
```

There are two weird things to note right away:

1) The colnames are the values from the table !
Solution: figure out more about read_tsv

```{r make empty cols}

# let's take a look at the documentation for read_tsv and see if there is a parameter we can use.
?read_tsv

# cool so there is a col_names = F we can use -- let's try it.
Pol2_peak1 <- read_tsv("/scratch/Shares/rinnclass/CLASS_2023/JR/CLASS_2023/CLASSES/03_Nextflow/00_Pol2_NF_CORE_RUN/results/bwa/mergedLibrary/macs/broadPeak/POLR2A_R1_peaks.broadPeak", col_names = F)

# Nice so now we have X1-X9 as col names
```

# we can make this more succinct by setting a basepath (you can call it anything)
# Then we can put them together with the rest of the path
```{r }

# change <your_folder> to your folder name
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/<your_folder>"

# now we can read in the file using file.path() function , rest of path
Pol2_peak1 <- read_tsv(file.path(basepath, "CLASS_2023/CLASSES/03_Nextflow/00_Pol2_NF_CORE_RUN/results/bwa/mergedLibrary/macs/broadPeak/POLR2A_R1_peaks.broadPeak"), col_names = F)

# note we had to shut off base.path function with closing ) then do col_names = F
```


2) There are weird chromosome names that are mislocated DNA "contigs" 
Solution: filter them out.

Let's first rename the cols. Well we need to know what they are :) 
So let's look up the labels of a .broad peak:

Chromosome, start, stop, name, score, strand, signalValue, pValue, qValue

# Signal value = overall enrichment of peak over input
# score is how dark the peak will show up in browser (not very useful :)

Ok now we are ready to rename the cols.
There are so many ways to do this we will just explore a couple
```{r renaming columns}

# first let's just change the name of one col (we have X1-X9)

names(Pol2_peak1) <- 'chromosome'

# ok changed the right col name, but now the other cols have no names !!
# not to worry we can index them and rename


# here we used names and then outside that indexed col 2 with [] 
names(Pol2_peak1)[2] <- 'start'

# This will get really repetitive but works. Here is a better solution:

names(Pol2_peak1) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue')

# here C makes a concatenated list of values -- trick is to make sure in right order!!
```

Cool so now we have a nice organized column names.

# Now let's filter out the weird chromosomes. 
# There are many ways to do this we will test a few

```{r filtering to cannonical chromosomes}

# We are going to first use some "indexing" we will do this A LOT !!
# we keep same object name, we use that object to index using []

Pol2_peak1 <- Pol2_peak1[Pol2_peak1$chromosome %in% c(paste0("chr", 1:22), "chrM", "chrX", "chrY"),]

# Note we have an empty , at the end -- this is because R reads as rows, cols 
# so the empty , is just saying keep all the cols and change the rows 
# also note the use of tidyverse %in% which is like filter -- item has to be "IN" this list.


# here we can do the same thing with less indexing using tidyverse "filter" function

Pol2_peak1 <- filter(Pol2_peak1, chromosome %in% c(paste0("chr", 1:22), "chrM", "chrX", "chrY"))

# tidyverse is awesome because it can now what is inside a data frame with out indexing.
# note we just say "chromosome" to call that col. Above more bulky Pol2_peak1$chromosome


# Let's get even more tidy:

Pol2_peak1 <- Pol2_peak1 %>%
  filter(chromosome %in% c(paste0("chr", 1:22), "chrM", "chrX", "chrY")) 

# Here we use the tidyverse pipe "%>%" to send the information in dataframe (DF)
# to filter and then add the parameters in filter

```

Nice, so now we have a data frame we can start to analyze

# Let's check out some basic features:

```{r basic summary of data frame}

# how many peaks are there -- well it's kinda obvious from env
# but we can use baseR to determine the same value:

length(Pol2_peak1)

# this tells us there are 9 cols

length(Pol2_peak1$chromosome)

# this tells us the chromosome col has 33,493 peaks.


# let's try using summary and table baseR functions;

summary(Pol2_peak1)

# Woah lots of information. Each column is summarized in quartiles etc
# we can see the Pvalues range from 10^-3.401 (0.0003971915) to 10^-154.989 (1.025652e-155)
# qvlaue min is 10^-1.770 (0.01698244) > so MACS uses a 0.01 qValue cut of for sig peaks


# Now let's try table to see how many times a qValue is the same

table(Pol2_peak1$qValue)

# yikes that is hard to understand. But we can table(table :)

table(table(Pol2_peak1$qValue))

# so we can see 1 qvalue shows up 229 times but most have a unique qValue

# let's see if fold change is as variable 
table(table(Pol2_peak1$signalValue))

```

Now let's make a new column that has the width of the peak using "mutate" function

```{r mutate}

# let's see what mutate needs
?mutate

# we will use tidyverse to pipe the DF to mutate so it knows col names

Pol2_peak1 <- Pol2_peak1 %>%
  # mutate wants name of new col and then function to create that col
  # we will simply subtract the end and start of the peak.
  
  mutate(peak_width = end - start)


# That is very handy and a function we will use a lot !!

```


Let's make some basic plots to get to know our data better:






